src/main/java/com/google/step/snippet/servlets/FeedbackServlet.java

```diff
@@ -0,0 +1,174 @@
+        if (userUpCards != null && userUpCards.contains(url)) {
+          json.put(STATUS, UPVOTED);
+        } else if (userDownCards != null && userDownCards.contains(url)) {
+          json.put(STATUS, DOWNVOTED);
+        }
+      }
+    }
+    response.setContentType("application/json");
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    // Invalid POST request if URL is invalid or user is not logged
+    String url = request.getParameter(URL);
+    UserService userService = UserServiceFactory.getUserService();
+    if (url == null || !userService.isUserLoggedIn()) {
+      JSONObject json = new JSONObject();
+      json.put(STATUS, NONE);
+      response.setContentType("application/json");
+      response.getWriter().println(json);
+      return;
+    }
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity userEntity = getUserEntity(datastore, userService);
+
+    // Get lists of user's upvoted and downvoted cards
+    ArrayList<String> upCards;
+    if (userEntity.getProperty(UPVOTED) == null) {
+      upCards = new ArrayList<String>();
+    } else {
+      upCards = (ArrayList<String>) userEntity.getProperty(UPVOTED);
+    }
+    ArrayList<String> downCards;
+    if (userEntity.getProperty(DOWNVOTED) == null) {
+      downCards = new ArrayList<String>();
+    } else {
+      downCards = (ArrayList<String>) userEntity.getProperty(DOWNVOTED);
+    }
+
+    // Add or remove card from user's list of upvoted and downvoted cards
+    if (request.getParameter(UP) != null) {
+      if (!upCards.contains(url)) {
+        upCards.add(url);
+        if (downCards.contains(url)) {
+          downCards.remove(url);
+        }
+      } else {
+        upCards.remove(url);
+      }
+    } else if (request.getParameter(DOWN) != null) {
+      if (!downCards.contains(url)) {
+        downCards.add(url);
+        if (upCards.contains(url)) {
+          upCards.remove(url);
+        }
+      } else {
+        downCards.remove(url);
+      }
+    }
+    userEntity.setProperty(UPVOTED, upCards);
+    userEntity.setProperty(DOWNVOTED, downCards);
+    datastore.put(userEntity);
+
+    // Retrieve toggle status of upvote and downvote buttons for user
+    String status = NONE;
+    if (userEntity.getProperty(UPVOTED) != null
+        && ((ArrayList<String>) userEntity.getProperty(UPVOTED)).contains(url)) {
+      status = UPVOTED;
+    }
+    if (userEntity.getProperty(DOWNVOTED) != null
+        && ((ArrayList<String>) userEntity.getProperty(DOWNVOTED)).contains(url)) {
+      status = DOWNVOTED;
+    }
+
+    // Calculate the total votes per card
+    long totalUpvotes = 0;
+    long totalDownvotes = 0;
+    Query allUsers = new Query(USER);
+    PreparedQuery results = datastore.prepare(allUsers);
+    for (Entity entity : results.asIterable()) {
```

Comment:
There's almost definitely a Datastore way to avoid this loop and do it right from the query (similar to how you can avoid doing a loop by adding a filter predicate), but neither of us have time to look into it. Just know that counting things satisfying a predicate is a basic thing SQL queries can do.